src/control/server/iosrv.go

```diff
@@ -148,3 +152,116 @@ func writeIosrvSuper(path string, super *iosrvSuper) error {
 
 	return common.WriteFileAtomic(path, data, 0600)
 }
+
+// iosrv represents an I/O server.
+//
+// NOTE: The superblock supercedes the format-time configuration in config.
+type iosrv struct {
+	super   *iosrvSuper
+	config  *configuration
+	index   int
+	cmd     *exec.Cmd
+	sigchld chan os.Signal
+	ready   chan struct{}
+	conn    *drpc.ClientConnection
+}
+
+func newIosrv(config *configuration, i int) (*iosrv, error) {
+	super, err := readIosrvSuper(iosrvSuperPath(config.Servers[i].ScmMount))
+	if err != nil {
+		return nil, err
+	}
+
+	srv := &iosrv{
+		super:   super,
+		config:  config,
+		index:   i,
+		sigchld: make(chan os.Signal, 1),
+		ready:   make(chan struct{}),
+		conn:    getDrpcClientConnection(config.SocketDir),
+	}
+
+	return srv, nil
+}
+
+func (srv *iosrv) start() (err error) {
+	defer func() { err = errors.WithMessagef(err, "start server %s", srv.config.Servers[srv.index].ScmMount) }()
+
+	if err = srv.startCmd(); err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			srv.stopCmd()
+		}
+	}()
+
+	// Wait for the notifyReady request or the SIGCHLD from the I/O server.
+	// If the I/O server is ready, make a dRPC connection to it.
+	select {
+	case <-srv.ready:
+	case <-srv.sigchld:
+		return errors.New("received SIGCHLD")
+	}
+	if err = srv.conn.Connect(); err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			srv.conn.Close()
+		}
+	}()
+
+	// TODO:
+	//   1 Get the rank and send a SetRank request to the I/O server.
+	//   2 If CreateMs, send a CreateMs request to the I/O server.
+	//   3 Clear CreateMs and BootstrapMs in the superblock.
+	//   4 Send a StartMs request to the I/O server.
+
+	return
+}
+
+func (srv *iosrv) wait() error {
+	if err := srv.cmd.Wait(); err != nil {
+		return errors.Wrapf(err, "wait server %s", srv.config.Servers[srv.index].ScmMount)
```

Comment:
so the rationale for using wrap here is to preserve the stack trace from external code?

Reply:
Here err doesn't have a stack trace, so we wrap it to add a message and a full stack trace.